contrib/release.sh

```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+#
+# Check that we can publish crates in their current form if there are changes on top of the tip of
+# master that imply that we are about to do a release.
+
+set -ex
+
+main () {
+    for crate in "internals" "hashes" "bitcoin"; do
+        if release_changes $crate; then
+            echo "$crate has changes implying this is a release PR"
+            publish_dry_run $crate
+        fi
+    done
+
+    exit 0
+}
+
+# Returns 0 if crate ($1) contains changes since tip of master that imply this patch set is done in
+# preparation for releasing the crate.
+release_changes() {
+    local crate=$1
+    if git log --patch --reverse master.. | grep --quiet "^\+\+\+ b/$crate/Cargo.toml"; then
+        if git log --patch --reverse master.. | grep --quiet "^\+version"; then
```

Comment:
This logic could lead to false positives if there are changes to one crate's manifest but _not_ the version field and the version field is updated in another manifest - does this matter? Someone want to take flex their shell-fu?

Reply:
You can filter `git log` by file, so you can one-line this (and then drop the `if` since it looks like you're just forwarding the return value anyway):
```
release_changes() {
    git log --patch --reverse master.. -- $crate/Cargo.toml | grep version || true
}
```
The `|| true` is to prevent `set -e` from killing the script when this fails. I'd also drop the `--quiet` from grep. If there are changes it might be nice to see exactly what they are.

Reply:
Oops, `|| true` will not forward the return value. Maybe we do need an `if` here.

Reply:
oooh, `-- <file>` is what I was looking for - nice! Will re-work

Reply:
Since, this is only for CI, you can also query crates.io to get the latest version. For the current PR, you can use env variables that cargo sets for you. See: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates

Reply:
No strong preference, just suggesting another alternative.

Reply:
Cool, cheers.

Reply:
On further thought, maybe we **don't** need the `if`, since we are calling `release_changes` in an `if`. I'm not sure.

Reply:
We don't seem to need the `|| true`. We now have:
```bash
release_changes() {
    local crate=$1
    git log --patch --reverse master.. -- $crate/Cargo.toml | grep version
}
```